File: src/scraper/uriMapper.ts
export function domainSlug(host: string): string {
  return host
    .toLowerCase()
    .replace(/^www\./, "")
    .replace(/[.:]/g, "-");
}

export function normalizeUrl(url: string): string {
  const parsed = new URL(url);
  parsed.hash = "";
  let result = parsed.origin + parsed.pathname;
  if (result.endsWith("/") && result !== parsed.origin + "/") {
    result = result.slice(0, -1);
  }
  // strip trailing slash even for origin
  return result.replace(/\/$/, "");
}

export function urlToUri(url: string, sectionHeading?: string): string {
  const parsed = new URL(url);
  const slug = domainSlug(parsed.host);
  const pathSegments = parsed.pathname
    .split("/")
    .filter(Boolean)
    .map((s) => encodeURIComponent(s.toLowerCase()));

  const base = ["contextfs://scraped", slug, ...pathSegments].join("/");

  if (sectionHeading) {
    const headingSlug = sectionHeading
      .toLowerCase()
      .replace(/[^a-z0-9]+/g, "-")
      .replace(/^-|-$/g, "");
    return `${base}/${headingSlug}`;
  }

  return base;
}

export function urlToParentUri(url: string): string | null {
  const uri = urlToUri(url);
  // contextfs://scraped/domain  →  null
  // contextfs://scraped/domain/a  →  contextfs://scraped/domain
  // contextfs://scraped/domain/a/b  →  contextfs://scraped/domain/a
  const parts = uri.split("/");
  // parts: ["contextfs:", "", "scraped", "domain", ...path]
  if (parts.length <= 4) return null; // only domain segment, no parent
  return parts.slice(0, -1).join("/");
}

File: mairu/internal/scraper/uri_mapper.go
package scraper

import (
	"net/url"
	"regexp"
	"strings"
)

const URIPrefix = "contextfs://scraped/"

func domainSlug(host string) string {
	host = strings.ToLower(host)
	host = strings.TrimPrefix(host, "www.")
	host = strings.ReplaceAll(host, ".", "-")
	host = strings.ReplaceAll(host, ":", "-")
	return host
}

func NormalizeURL(rawURL string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return rawURL
	}
	parsed.Fragment = ""
	res := parsed.Scheme + "://" + parsed.Host + parsed.Path
	if strings.HasSuffix(res, "/") && res != parsed.Scheme+"://"+parsed.Host+"/" {
		res = res[:len(res)-1]
	}
	return strings.TrimSuffix(res, "/")
}

var nonAlnumRegex = regexp.MustCompile(`[^a-z0-9]+`)

func URLToURI(rawURL string, sectionHeading ...string) string {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return URIPrefix + "invalid"
	}
	slug := domainSlug(parsed.Host)

	parts := strings.Split(parsed.Path, "/")
	var validParts []string
	for _, p := range parts {
		if p != "" {
			validParts = append(validParts, url.QueryEscape(strings.ToLower(p)))
		}
	}

	base := URIPrefix + slug
	if len(validParts) > 0 {
		base += "/" + strings.Join(validParts, "/")
	}

	if len(sectionHeading) > 0 && sectionHeading[0] != "" {
		headingSlug := strings.ToLower(sectionHeading[0])
		headingSlug = nonAlnumRegex.ReplaceAllString(headingSlug, "-")
		headingSlug = strings.Trim(headingSlug, "-")
		return base + "/" + headingSlug
	}

	return base
}

func URLToParentURI(rawURL string) *string {
	uri := URLToURI(rawURL)
	parts := strings.Split(uri, "/")
	if len(parts) <= 4 {
		return nil
	}
	res := strings.Join(parts[:len(parts)-1], "/")
	return &res
}
